In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

# **DATASET INTO DATAFRAME**

In [2]:
UK = pd.read_csv('filedata\\uk_road_accident.csv')

In [3]:
UK

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [4]:
UK.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [5]:
UK.describe().T

count       mean       std        min        25%  \
Latitude              660654.0  52.553866  1.406922  49.914430  51.490690   
Longitude             660653.0  -1.431210  1.383330  -7.516225  -2.332291   
Number_of_Casualties  660679.0   1.357040  0.824847   1.000000   1.000000   
Number_of_Vehicles    660679.0   1.831255  0.715269   1.000000   1.000000   

                            50%        75%        max  
Latitude              52.315641  53.453452  60.757544  
Longitude             -1.411667  -0.232869   1.762010  
Number_of_Casualties   1.000000   1.000000  68.000000  
Number_of_Vehicles     2.000000   2.000000  32.000000

In [6]:
UK.isnull()

Index  Accident_Severity  Accident Date  Latitude  Light_Conditions  \
0       False              False          False     False             False   
1       False              False          False     False             False   
2       False              False          False     False             False   
3       False              False          False     False             False   
4       False              False          False     False             False   
...       ...                ...            ...       ...               ...   
660674  False              False          False     False             False   
660675  False              False          False     False             False   
660676  False              False          False     False             False   
660677  False              False          False     False             False   
660678  False              False          False     False             False   

        District Area  Longitude  Number_of_Casualties  Number_of_Vehicles  \
0               False      False                 False               False   
1               False      False                 False               False   
2               False      False                 False               False   
3               False      False                 False               False   
4               False      False                 False               False   
...               ...        ...                   ...                 ...   
660674          False      False                 False               False   
660675          False      False                 False               False   
660676          False      False                 False               False   
660677          False      False                 False               False   
660678          False      False                 False               False   

        Road_Surface_Conditions  Road_Type  Urban_or_Rural_Area  \
0                         False      False                False   
1                         False      False                False   
2                         False       True                False   
3                         False      False                False   
4                         False       True                False   
...                         ...        ...                  ...   
660674                    False      False                False   
660675                    False      False                False   
660676                    False      False                False   
660677                    False      False                False   
660678                    False      False                False   

        Weather_Conditions  Vehicle_Type  
0                    False         False  
1                    False         False  
2                     True         False  
3                    False         False  
4                     True         False  
...                    ...           ...  
660674               False         False  
660675               False         False  
660676               False         False  
660677               False         False  
660678               False         False  

[660679 rows x 14 columns]

In [7]:
UK.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

# **SELECTING THE COLUMNS**

In [8]:
UK['Accident_Severity']

0         Serious
1         Serious
2         Serious
3         Serious
4          Slight
           ...   
660674     Slight
660675     Slight
660676     Slight
660677    Serious
660678    Serious
Name: Accident_Severity, Length: 660679, dtype: object

In [9]:
UK['Accident_Severity'] == 'Serious'

0          True
1          True
2          True
3          True
4         False
          ...  
660674    False
660675    False
660676    False
660677     True
660678     True
Name: Accident_Severity, Length: 660679, dtype: bool

In [10]:
UK['Accident Date']

0           5/6/2019
1           2/7/2019
2         26-08-2019
3         16-08-2019
4           3/9/2019
             ...    
660674    18-02-2022
660675    21-02-2022
660676    23-02-2022
660677    23-02-2022
660678    28-02-2022
Name: Accident Date, Length: 660679, dtype: object

In [11]:
UK['Accident Date'] == '26-08-2019'

0         False
1         False
2          True
3         False
4         False
          ...  
660674    False
660675    False
660676    False
660677    False
660678    False
Name: Accident Date, Length: 660679, dtype: bool

In [12]:
UK['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

In [13]:
UK['Accident_Severity'].unique()

array(['Serious', 'Slight', 'Fatal'], dtype=object)

In [14]:
Fatal_df = UK[UK['Accident_Severity'] == 'Fatal']

In [15]:
Fatal_df

Index Accident_Severity Accident Date   Latitude  \
7       200701BS67370             Fatal     3/10/2019  51.482260   
11      200701BS67747             Fatal     2/11/2019  51.499127   
20      200701BS68301             Fatal    24-10-2019  51.505996   
21      200701BS68302             Fatal    24-11-2019  51.486795   
717     200701CP00032             Fatal    30-01-2019  51.510978   
...               ...               ...           ...        ...   
660470  201091NH00529             Fatal    29-01-2022  58.475726   
660519  201091NJ01533             Fatal      2/2/2022  58.495554   
660565  201091NJ11329             Fatal    15-08-2022  57.871514   
660581  201091NJ14695             Fatal    21-10-2022  58.445475   
660627  201091NL00060             Fatal      6/1/2022  60.241195   

             Light_Conditions           District Area  Longitude  \
7       Darkness - lights lit  Kensington and Chelsea  -0.179486   
11                   Daylight  Kensington and Chelsea  -0.159652   
20                   Daylight  Kensington and Chelsea  -0.214277   
21                   Daylight  Kensington and Chelsea  -0.170375   
717                  Daylight          City of London  -0.107581   
...                       ...                     ...        ...   
660470               Daylight           Western Isles  -6.278583   
660519               Daylight                Highland  -4.451628   
660565               Daylight                Highland  -4.243081   
660581  Darkness - lights lit                Highland  -3.065535   
660627               Daylight        Shetland Islands  -1.409761   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
7                          3                   2                     Dry   
11                         1                   2                     Dry   
20                         1                   1                     Dry   
21                         5                   1                     Dry   
717                        1                   2             Wet or damp   
...                      ...                 ...                     ...   
660470                     1                   1             Wet or damp   
660519                     1                   1                    Snow   
660565                     1                   1                     Dry   
660581                     1                   1             Wet or damp   
660627                     3                   1                    Snow   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
7       Single carriageway               Urban     Fine no high winds   
11      Single carriageway               Urban     Fine no high winds   
20      Single carriageway               Urban     Fine no high winds   
21      Single carriageway               Urban     Fine no high winds   
717       Dual carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660470  Single carriageway               Rural      Fine + high winds   
660519  Single carriageway               Rural  Snowing no high winds   
660565  Single carriageway               Rural     Fine no high winds   
660581  Single carriageway               Rural     Fine no high winds   
660627  Single carriageway               Rural     Fine no high winds   

                                Vehicle_Type  
7                                        Car  
11      Bus or coach (17 or more pass seats)  
20                                       Car  
21                                       Car  
717                                      Car  
...                                      ...  
660470                                   Car  
660519                                   Car  
660565         Goods 7.5 tonnes mgw and over  
660581                                   Car  
660627   Van / Goods 3.5 tonnes mgw or under  

[8661 rows x 14 columns]

In [16]:
Serious_df = UK[UK['Accident_Severity'] == 'Serious']

In [17]:
Slight_df = UK[UK['Accident_Severity'] == 'Slight']

In [18]:
UK[UK['Latitude'] >= 50]

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [19]:
UK[(UK['Accident_Severity'] == "Fatal") & (UK['Latitude'] >= 50)].info()

<class 'pandas.core.frame.DataFrame'>
Index: 8661 entries, 7 to 660627
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Index                    8661 non-null   object 
 1   Accident_Severity        8661 non-null   object 
 2   Accident Date            8661 non-null   object 
 3   Latitude                 8661 non-null   float64
 4   Light_Conditions         8661 non-null   object 
 5   District Area            8661 non-null   object 
 6   Longitude                8661 non-null   float64
 7   Number_of_Casualties     8661 non-null   int64  
 8   Number_of_Vehicles       8661 non-null   int64  
 9   Road_Surface_Conditions  8659 non-null   object 
 10  Road_Type                8628 non-null   object 
 11  Urban_or_Rural_Area      8661 non-null   object 
 12  Weather_Conditions       8554 non-null   object 
 13  Vehicle_Type             8661 non-null   object 
dtypes: float64(2), int64(2), ob

# FILL UP NULL VALUES

In [20]:
UK.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [21]:
UK['Latitude'].mean()

np.float64(52.553865761110956)

In [22]:
UK['Latitude'].fillna(UK['Latitude'].mean())

0         51.506187
1         51.495029
2         51.517715
3         51.495478
4         51.488576
            ...    
660674    57.374005
660675    57.232273
660676    57.585044
660677    57.214898
660678    57.575210
Name: Latitude, Length: 660679, dtype: float64

In [ ]:
UK['Latitude'] = UK['Latitude'].fillna(UK['Latitude'].mean())
UK['Longitude'] = UK['Longitude'].fillna(UK['Longitude'].mean())
UK['Index'] = UK['Index'].fillna(UK['Index'].mean())

In [24]:
print(UK.columns)

Index(['Index', 'Accident_Severity', 'Accident Date', 'Latitude',
       'Light_Conditions', 'District Area', 'Longitude',
       'Number_of_Casualties', 'Number_of_Vehicles', 'Road_Surface_Conditions',
       'Road_Type', 'Urban_or_Rural_Area', 'Weather_Conditions',
       'Vehicle_Type'],
      dtype='object')


# **5 Insights from the First Look at the Dataset** (DATAFRAME)

## 1. **What do the first rows of the dataset show?** 
###   *-The first few rows display information about different accidents, including severity, date, location, and vehicle details.*

## 2. **What kind of details are recorded?** 
###  *-Each row has information such as accident severity, number of vehicles, number of casualties, weather, and road conditions.*  

## 3. **Are there location details in the dataset?**  
###  *-Yes, the dataset includes latitude and longitude for each accident, which means accidents can be mapped geographically.* 

## 4. **Do the accident records show dates?** 
###  *-Yes, the dataset includes an “Accident Date” column, showing when each accident happened.* 

## 5. **Is vehicle information part of the data?** 
###  *-Yes, there is a column called “Vehicle_Type” which shows the type of vehicle involved in the accidents.*  


# 4 **Insights from the Dataset** (.info())
## 1. **How many rows are in the dataset?**  
###  *-The dataset has 660,679 rows which means it records over six hundred thousand road accidents.*  

## 2. **How many columns are included?**  
###  *-There are 14 columns that describe details like accident severity, vehicles, casualties, weather, and so on.*  

## 3. **What types of data are in the columns?**  
###  *-Most of the columns are in object type (text), while only a few like Number_of_Casualties, Number_of_Vehicles, Latitude, and Longitude are numeric.*  

## 4. **Are there any missing values?**  
###  *-Yes, some columns have missing values, for example Weather_Conditions is missing a lot of data, and latitude/longitude are missing a few.*  


# **3 Insights from the Summary Statistics** (.describe())

## 1. **What does the average number of casualties show?**  
###  *-On average, there is about 1 casualty per accident, which means most accidents only involve a small number of people.*  

## 2. **What about the number of vehicles per accident?**  
###  *-The mean number of vehicles is around 1 to 2, showing that many accidents involve just one or two vehicles.*  

## 3. **Do latitude and longitude values look consistent?**  
###  *-Yes, the latitude and longitude ranges are close together without extreme outliers, meaning accident locations are recorded properly across the UK.*  


# **6 Insights from Checking Missing Values** (.isnull())

## 1. **Does the dataset have any missing values overall?**  
###  *-Yes, the dataset is not fully complete — some columns have missing information.*  

## 2. **Which columns are almost complete?**  
###  *-Columns like Accident_Severity, Accident Date, District Area, and Vehicle_Type have no missing values.*  

## 3. **What about latitude and longitude?**  
###  *-Latitude is missing 25 entries, and Longitude is missing 26 entries out of more than 660,000 — a very small amount compared to the whole dataset.*  

## 4. **Which column has the largest number of missing values?**  
###  *-Weather_Conditions has the most missing data, with over 14,000 entries left blank.*  

## 5. **Are road details also missing in some cases?**  
###  *-Yes, Road_Type and Road_Surface_Conditions have a few thousand missing records.*  

## 6. **What does this mean for analysis?**  
###  *-Most of the dataset is complete, but some analyses (like weather impact or road type studies) might need missing values to be filled or ignored.*  


# **2 Insights from Accident Severity** (UK['Accident_Severity'])

## 1. **What does the Accident_Severity column contain?**  
###  *-The column shows the level of each accident, such as Slight, Serious, or Fatal.*  

## 2. **What happens when we check if Accident_Severity equals ‘Serious’?**  
###  *-It gives a list of True/False values, where True means the accident was serious and False means it was not.*  


# **3 Insights from Accident Date**

## 1. **What kind of data is in the Accident Date column?**  
###  *-It contains the dates when each accident happened, written in day-month-year format.*  

## 2. **Can we check for a specific date?**  
###  *-Yes, by using a condition like UK['Accident Date'] == '26-08-2019', we can see which rows belong to that exact date.*  

## 3. **What does the condition return?**  
###  *-It gives a series of True/False values, where True means the accident happened on 26-08-2019 and False means it did not.*  


# **3 Insights from Accident Severity Counts**

## 1. **Which type of accidents are the most common?**  
###  *-The majority of accidents are labeled as **Slight**, making up the largest portion of the dataset.*  

## 2. **How many serious accidents are there?**  
###  *-There are thousands of accidents marked as **Serious**, showing that a significant number of cases are more severe than slight accidents.*  

## 3. **Are fatal accidents common?**  
###  *-Fatal accidents are the least common compared to Slight and Serious, but they are still an important part of the data for safety analysis.*  


# **10 Insights from Accident Severity and Fatal Accidents**

## 1. **What are the unique accident severity levels?**  
###  *-There are three categories: Slight, Serious, and Fatal.*  

## 2. **Why is this important?**  
###  *-It shows that all accidents in the dataset can be grouped into these three levels of severity.*  

## 3. **What does Fatal_df contain?**  
###  *-Fatal_df is a new dataset created by filtering only the accidents marked as Fatal.*  

## 4. **How many rows are in Fatal_df?**  
###  *-It contains only a small portion of the total accidents, since fatal cases are less frequent than others.*  

## 5. **Do Fatal accidents have the same columns?**  
###  *-Yes, Fatal_df still keeps all 14 columns, but only with rows where Accident_Severity = Fatal.*  

## 6. **Why is filtering useful?**  
###  *-Filtering allows us to focus only on the most serious type of accidents without mixing them with other data.*  

## 7. **Can we compare Fatal accidents with others?**  
###  *-Yes, by comparing Fatal_df with Serious or Slight accidents, we can study differences in conditions or causes.*  

## 8. **Does Fatal_df also have missing values?**  
###  *-Yes, it still has missing values in some columns like Weather_Conditions or Road_Type, just like the full dataset.*  

## 9. **How can Fatal_df help analysis?**  
###  *-It helps researchers and policymakers concentrate on the most dangerous accidents for better road safety planning.*  

## 10. **What is the benefit of using unique() before filtering?**  
###  *-By knowing the categories (Slight, Serious, Fatal), we can easily choose which one to filter, such as creating Fatal_df.*  


# **5 Insights from Filtering Latitude and Fatal Accidents**

## 1. **What does Latitude >= 50 select?**  
###  *-It keeps all accidents that happened in locations at latitude 50 or higher, which covers a big part of the UK.*  

## 2. **What happens when we combine Latitude >= 50 with Fatal accidents?**  
###  *-The filter shows only the most severe accidents (Fatal) that occurred in the same latitude range.*  

## 3. **What does .info() tell us about this filtered data?**  
###  *-It shows the number of rows that match the filter, along with the details of all 14 columns.*  

## 4. **Does the filtered dataset still include missing values?**  
###  *-Yes, even within this smaller set, some columns like Weather_Conditions or Road_Type can still have missing values.*  

## 5. **Why is combining filters useful?**  
###  *-By combining Accident Severity with Latitude, we can study where the most dangerous accidents happen in specific parts of the UK.*  
